### PR TITLE
fix: able to add repo name with "."(dot) in it issue#1451

### LINKED ISSUE #1451: Bug: When adding brob/plug11ty.com "not valid" error
### Describe the bug

When adding a repository from my synced GitHub account, I get the message "Invalid GitHub repository"

### Steps to reproduce

1. Go to https://app.opensauced.pizza/
2. In `owner/repo` field type brob/plug11ty.com (corresponds to [this repo](https://github.com/brob/plug11ty.com))
3. Click "Add"
4. Error message "Invalid GitHub repository" appears

### Affected services

opensauced.pizza

### Platforms

Desktop

### Browsers

Chrome

### Environment

Production

### Additional context

_No response_

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### Contributing Docs

- [X] I agree to follow this project's Contribution Docs

### PR DESCRIPTION
**Pull request by arth2002**
## What type of PR is this? (check all applicable)

- [ ] 🍕 Feature
- [x] 🐛 Bug Fix
- [ ] 📝 Documentation Update
- [ ] 🎨 Style
- [ ] 🧑‍💻 Code Refactor
- [ ] 🔥 Performance Improvements
- [ ] ✅ Test
- [ ] 🤖 Build
- [ ] 🔁 CI
- [ ] 📦 Chore (Release)
- [ ] ⏩ Revert

## Description

This PR fix #1451 

## Mobile & Desktop Screensh
Screenshot from 2022-06-19 16-41-26
ots/Recordings

## Added tests?

- [ ] 👍 yes
- [x] 🙅 no, because they aren't needed
- [ ] 🙋 no, because I need help

## Added to documentation?

- [ ] 📜 README.md
- [ ] 📓 docs.opensauced.pizza
- [ ] 🍕 dev.to/opensauced
- [ ] 📕 storybook
- [ ] 🙅 no documentation needed

### CODE REVIEW

**Comment by mtfoley** [comment]
@arth2002 thanks for getting this PR started.

@0-vortex with that failing test I guess if what we really care about are the ones that start with github.com, we could make that a third case. WDYT?

**Comment by 0-vortex** [comment]
> @arth2002 thanks for getting this PR started.
> 
> @0-vortex with that failing test I guess if what we really care about are the ones that start with github.com, we could make that a third case. WDYT?

We should support all repos, this should fix things not become a breaking change. Changing from `includes->substr` will do the trick - it's a "relative" URL only if it starts with those matchers, not if it's included in `repoName`

**Comment by mtfoley** [comment]
I checked out this PR locally, and I actually think the utils.js would work well like this:
```js
export function isValidRepoUrl(url) {
  
  return fullnameValidator(urlToFullName(url));
}

function urlToFullName(url){
  const fullname = url.replace(/^(http|https):\/\//i, "") // take off URL protocol
    .replace(/^(www\.)*github\.com\//i, "") // take off github.com variant
    .replace(/^\/*/ ,"")// take off leading slash(es) if there
    .replace(/ /g, ""); // take out spaces
  return fullname;
}

function fullnameValidator(fullname){
  // take only first two elements
  const [owner, repo] = fullname.split("/");
  if(!owner || !repo) return [false, null];
  // alphanumerics and hyphens are allowed in owner names
  const invalidOwnerRegex = /[^a-z0-9-]+/i;
  // alphanumerics, hyphens, underscores, and dots are allowed in repo names
  const invalidRepoRegex = /[^a-z0-9-_\.]+/i;
  const valid = !invalidOwnerRegex.test(owner) && !invalidRepoRegex.test(repo);
  // if invalid, return [false, null]
  // if valid, return [true, fullname]
  return [valid, valid ? fullname : null];
}

```

@arth2002 I think the code above needs some more tweaking but I wanted to bounce the idea off of you and @0-vortex.

The test file actually has some issues as well so I'll follow up with a later comment.

**Comment by 0-vortex** [comment]
> I checked out this PR locally, and I actually think the utils.js would work well like this:
> 
> ```js
> export function isValidRepoUrl(url) {
>   
>   return fullnameValidator(urlToFullName(url));
> }
> 
> function urlToFullName(url){
>   const fullname = url.replace(/^(http|https):\/\//i, "") // take off URL protocol
>     .replace(/^(www\.)*github\.com\//i, "") // take off github.com variant
>     .replace(/^\/*/ ,"")// take off leading slash(es) if there
>     .replace(/ /g, ""); // take out spaces
>   return fullname;
> }
> 
> function fullnameValidator(fullname){
>   // take only first two elements
>   const [owner, repo] = fullname.split("/");
>   if(!owner || !repo) return [false, null];
>   // alphanumerics and hyphens are allowed in owner names
>   const invalidOwnerRegex = /[^a-z0-9-]+/i;
>   // alphanumerics, hyphens, underscores, and dots are allowed in repo names
>   const invalidRepoRegex = /[^a-z0-9-_\.]+/i;
>   const valid = !invalidOwnerRegex.test(owner) && !invalidRepoRegex.test(repo);
>   // if invalid, return [false, null]
>   // if valid, return [true, fullname]
>   return [valid, valid ? fullname : null];
> }
> ```
> 
> @arth2002 I think the code above needs some more tweaking but I wanted to bounce the idea off of you and @0-vortex.
> 
> The test file actually has some issues as well so I'll follow up with a later comment.

Hard no on this - you have taken a function that checks whether a subset of the URL RFC is relative or not and turned it into a URL validator, that's potentially the opposite of what we are trying to do here; the regexp approach sounds good in theory but if you take a look at [RFC 1738](https://datatracker.ietf.org/doc/html/rfc1738) you will understand why this approach is so bad. You can spend weeks perfecting regular expressions like [this](https://rgxdb.com/r/2MQXJD5).

To make a better example, I mentioned in the previous review that the third `includes` should be changed into a `substr`, that is, essentially, to check for that condition at the beginning of the string. If you use regular expressions that can be simplified into `^(https?:\/\/)*(www\.?)*([a-z]+\.[a-z]+\/)*`, checking for all the parts of a URL and excluding them as they appear you get the problem that it has to be a perfect regex, all the parts have to work in all cases, which makes the difficulty of adjusting to all those use cases, terrible. The above regex would cover the following use cases:

```
brob/plug11ty.com
github.com/brob/plug11ty.com
www.github.com/brob/plug11ty.com
https://github.com/brob/plug11ty.com
https://www.github.com/brob/plug11ty.com
```

However, that is going full-on into scope creep. The existing code is meant to catch 3 things at the beginning of the string:
- http (done with `substr`)
- https (done with `substr`)
- www (done with regex without `^`)

Add the `^` to the third check or transform the regex into a `substr`

**Comment by mtfoley** [comment]
Fair enough.  @arth2002 is the instruction clear from what @0-vortex was saying?

**Comment by arth2002** [comment]
> Fair enough. @arth2002 is the instruction clear from what @0-vortex was saying?

Hey @mtfoley sorry to bother you sir
I just want to check we're on the same page,  as per I understood do I need to replace
this code
```
const isRelativeUrl = !(url.substr(0, 4) === "http" || url.includes(".com") || url.includes("www."));
```
with this one
```
 const isRelativeUrl = !(url.substr(0,7) === 'http://' || url.substr(0,8) === 'https://' || url.substr(0,4) === 'www.');
```

am I correct? if not can please correct me
Thank You :)

**Comment by arth2002** [comment]
Please review it

File: src/lib/util.js [hunk 1]
```diff
@@ -2,7 +2,7 @@ export function isValidRepoUrl(url) {
   url = url.trim();
```
Comment (mtfoley) [inline src/lib/util.js:2]: i think if you change this to `url = url.toLowerCase().trim();`, the tests will pass (they pass for me locally).

**Review by mtfoley — COMMENTED** [review]
@arth2002, I think it would also be helpful to add a couple of tests. The two cases I'd suggest would both be "valid", and would be:
1) the repo of the original issue posting ("brob/plug11ty.com")
2) the repo "httpie/httpie", just to verify that use case works also.

File: src/lib/util.js (hunk 1 above)
Comment (arth2002) [reply src/lib/util.js]: Done :thumbsup:
:crossed_fingers:(finger crossed)

File: src/lib/util.js [hunk 2]
```diff
@@ -1,8 +1,8 @@
 export function isValidRepoUrl(url) {
-  url = url.trim();
+  url = url.toLowerCase().trim();
   url = url.substr(0, 1) === "/" ? url.substr(1) : url;
```
Comment (mtfoley) [inline src/lib/util.js:3]: @arth2002 I just want you to know you're being very patient with me! That said, I'd suggest we take a moment to change `url.substr()` to be `url.substring()`. It looks like `substr()` is deprecated (though widely supported). https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/substr#browser_compatibility

Comment (arth2002) [reply src/lib/util.js]: Done :+1:

**Comment by mtfoley** [comment]
Looks good to me!

**Pull request merged**